hw/xbox/xid-sdl.c

```diff
@@ -464,6 +446,81 @@ static void usb_xid_class_initfn(ObjectClass *klass, void *data)
     uc->handle_attach  = usb_desc_attach;
 }
 
+static int apply_haptic_effect(USBXIDState *s, SDL_HapticEffect *effect) {
+    int effect_id = SDL_HapticNewEffect(s->sdl_haptic, effect);
+    if (effect_id == -1) {
+        const char* error = SDL_GetError();
+        fprintf(stderr, "Failed to create haptic effect! SDL Error: %s\n", error);
+        SDL_HapticClose(s->sdl_haptic);
+        s->sdl_haptic = NULL;
+        s->sdl_haptic_effect = NULL;
+        return -1;
+    }
+
+    int run_result = SDL_HapticRunEffect(s->sdl_haptic, effect_id, 1);
+    if (run_result == -1) {
+        const char* error = SDL_GetError();
+        fprintf(stderr, "Failed to run haptic effect! SDL Error: %s\n", error);
+        SDL_HapticClose(s->sdl_haptic);
+        s->sdl_haptic = NULL;
+        s->sdl_haptic_effect = NULL;
+        return -1;
+    }
+
+    s->sdl_haptic_effect = effect;
+    s->sdl_haptic_effect_id = effect_id;
+    return 0;
+}
+
+static int haptic_effect_init(USBXIDState *s, SDL_HapticEffect *effect) {
+    if (SDL_HapticQuery(s->sdl_haptic) & SDL_HAPTIC_LEFTRIGHT) {
+        effect->leftright = (SDL_HapticLeftRight){
+            .type = SDL_HAPTIC_LEFTRIGHT,
+            .length = SDL_HAPTIC_INFINITY,
+            .large_magnitude = 0x0000,
+            .small_magnitude = 0x0000,
+        };
+
+        return apply_haptic_effect(s, effect);
+
+    } else if (SDL_HapticQuery(s->sdl_haptic) & SDL_HAPTIC_SINE) {
+        fprintf(stderr, "Joystick doesn't support left/right haptic feedback, emulating with SINE");
+
+        effect->periodic = (SDL_HapticPeriodic){
+            .type = SDL_HAPTIC_SINE,
+            .period = 1000,
+            .magnitude = 0,
+            .length = SDL_HAPTIC_INFINITY,
+            .attack_length = 0,
+            .fade_length = 0,
+            .delay = 0,
+        };
```

Comment:
If we are using sine, the frequency (`period`?) should probably be influenced by the left/right strength because those motors *do* run different frequencies.

Due to the added complexity, I'd probably avoid adding this fallback for now (split it into a second commit possibly, so we can apply it later if necessary). I'd assume that almost every device will support Left/Right anyway.

Reply:
My assumption was that sine rumble will never truly be fully "accurate" and is just an approximation for users with older controllers. The period here is taken from SDL's implementation of `SDL_HapticRumble`: https://github.com/spurious/SDL-mirror/blob/master/src/haptic/SDL_haptic.c#L773. Testing on my local controller feels fine with this mode.

I warn the user with a print to stderr when initializing the fallback, maybe I can make it more clear that the fallback implementation isn't considered accurate?

I guess that goes back to the xqemu's philosophy though, should we even have this implementation if it will never truly be accurate? I have no problem removing the code here if that is preferred.

Reply:
Of course it's an approximation, but imagine the following cases:

- 100% of the large motor, 0% small motor: ~ sine at amplitude 100 at a very specific period (low frequency, because large weight spins slow).
- 100% of the small motor, 0% large motor: ~ sine at amplitude 100 at a very specific period (high frequency, because small weight spins fast).

The current fallback would always use the same frequency - So even approximations can be good or bad.

Yes, we should still have it, *if it's necessary*. However, I'd assume that most devices will have Left/Right anyway for XInput, possibly using similar approximation as implied here to drive a different combination of motors (to be compatible with most games).

So while we should ensure our code can possibly handle it in the future, we shouldn't have to worry about it for now (it's not a problem yet, so why spend time solving it?).

If we encounter many broken devices or XInput changes / new force feeback standards arrive, then we might have to revisit this.
And if we do, we'll probably want it as accurate as possible (not necessarily because of philosophy, but simply because users will expect it).

Reply:
All very fair points. Thanks for spending the time for the breakdown.